pkg/prometheus/promcfg.go

```diff
@@ -272,6 +294,248 @@ func (cg *configGenerator) generateConfig(
+			})
+		} else if len(nsel.MatchNames) > 0 {
+			relabelings = append(relabelings, yaml.MapSlice{
+				{Key: "action", Value: "keep"},
+				{Key: "source_labels", Value: []string{"__meta_kubernetes_namespace"}},
+				{Key: "regex", Value: strings.Join(nsel.MatchNames, "|")},
+			})
+		}
+	}
+
+	// Filter targets based on correct port for the endpoint.
+	if ep.Port != "" {
+		relabelings = append(relabelings, yaml.MapSlice{
+			{Key: "action", Value: "keep"},
+			{Key: "source_labels", Value: []string{"__meta_kubernetes_pod_container_port_name"}},
+			{Key: "regex", Value: ep.Port},
+		})
+	} else if ep.TargetPort != nil {
+		if ep.TargetPort.StrVal != "" {
+			relabelings = append(relabelings, yaml.MapSlice{
+				{Key: "action", Value: "keep"},
+				{Key: "source_labels", Value: []string{"__meta_kubernetes_pod_container_port_name"}},
+				{Key: "regex", Value: ep.TargetPort.String()},
+			})
+		} else if ep.TargetPort.IntVal != 0 {
+			relabelings = append(relabelings, yaml.MapSlice{
+				{Key: "action", Value: "keep"},
+				{Key: "source_labels", Value: []string{"__meta_kubernetes_pod_container_port_number"}},
+				{Key: "regex", Value: ep.TargetPort.String()},
+			})
+		}
+	}
+
+	// Relabel namespace and pod and service labels into proper labels.
+	relabelings = append(relabelings, []yaml.MapSlice{
+		{
+			{Key: "source_labels", Value: []string{"__meta_kubernetes_pod_node_name"}},
+			{Key: "target_label", Value: "node"},
+		},
+		{
+			{Key: "source_labels", Value: []string{"__meta_kubernetes_namespace"}},
+			{Key: "target_label", Value: "namespace"},
+		},
+		{
+			{Key: "source_labels", Value: []string{"__meta_kubernetes_pod_container_name"}},
+			{Key: "target_label", Value: "container"},
+		},
+		{
+			{Key: "source_labels", Value: []string{"__meta_kubernetes_pod_name"}},
+			{Key: "target_label", Value: "pod"},
+		},
+		{
+			{Key: "source_labels", Value: []string{"__meta_kubernetes_pod_phase"}},
+			{Key: "target_label", Value: "pod_phase"},
+		},
+		{
+			{Key: "source_labels", Value: []string{"__meta_kubernetes_pod_ready"}},
+			{Key: "target_label", Value: "pod_ready"},
+		},
+	}...)
+
+	// Relabel targetLabels from Pod onto target.
+	for _, l := range m.Spec.PodTargetLabels {
+		relabelings = append(relabelings, yaml.MapSlice{
+			{Key: "source_labels", Value: []string{"__meta_kubernetes_pod_label_" + sanitizeLabelName(l)}},
+			{Key: "target_label", Value: sanitizeLabelName(l)},
+			{Key: "regex", Value: "(.+)"},
+			{Key: "replacement", Value: "${1}"},
+		})
+	}
+
+	// By default, generate a safe job name from the pod name.  We also keep
+	// this around if a jobLabel is set in case the targets don't actually have a
+	// value for it. A single pod may potentially have multiple metrics
+	// endpoints, therefore the endpoints labels is filled with the ports name or
+	// as a fallback the port number.
+
+	relabelings = append(relabelings, yaml.MapSlice{
+		{Key: "source_labels", Value: []string{"__meta_kubernetes_pod_name"}},
```

Comment:
I'm not sure this is a great default, job is intended to group workloads of the same type, this will cause distinct job labels for those. Maybe the generate name would be more appropriate? (not sure there is a meta label for those though)

Reply:
Isn't this also true services if the LabelSelector of a ServiceMonitorSpec applies to multiple and differently names services? 

Which generated name did you have in mind? It doesn't look like theres a good pod meta label to solve this (1). 
---
(1) https://prometheus.io/docs/prometheus/latest/configuration/configuration/#pod

Reply:
Most pods are not created directly, and their controller sets the `.metadata.generateName` which is the "scheme" the pods name was chosen, often good enough for grouping, but probably a bit too magic for this. I'd say maybe let's just default to the PodMonitor's name as the job label?

Reply:
Thanks for the explanation. 
Defaulting the job name to the PodMonitors key (`$namespace/$name`) instead.